### ISSUE TITLE
"copy" module fails on "--diff" if source file containes UTF-8 chars

### ISSUE THREAD

**Issue by Farzy** [ansible/ansible-modules-core#3730]
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

copy
##### ANSIBLE VERSION

```
ansible 2.0.2.0
  config file = /Users/ffarid/src/self-config/ansible.cfg
  configured module search path = Default w/o overrides
```
##### OS / ENVIRONMENT

Ansible OS : Mac OSX 10.11.4
Ansible is installed via "brew"
Target OS : Debian 8
##### SUMMARY

Using "copy" module with the "--diff" option on a file with UTF-8 chars in it produces the following error:

```
 [WARNING]: Failure when attempting to use callback plugin (<ansible.plugins.callback.minimal.CallbackModule object at 0x105455f10>):
'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)
```
##### STEPS TO REPRODUCE

```
echo "éé" > /tmp/accents.txt
ansible new -m copy -a "src=/tmp/accents.txt dest=/tmp/accents.txt" --diff
```

##### EXPECTED RESULTS

```
--- before
+++ after: /tmp/accents.txt
@@ -0,0 +1 @@
+éé
```
##### ACTUAL RESULTS

```
 [WARNING]: Failure when attempting to use callback plugin (<ansible.plugins.callback.minimal.CallbackModule object at 0x105455f10>):
'ascii' codec can't decode byte 0xc3 in position 1: ordinal not in range(128)

--- before
+++ after: /tmp/accents.txt
@@ -0,0 +1 @@
+éé
```

With "-vvvv" the output is preceded by:

```
Loaded callback minimal of type stdout, v2.0
<silence.farzy.org> ESTABLISH SSH CONNECTION FOR USER: root
<silence.farzy.org> SSH: EXEC ssh -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r silence.farzy.org '/bin/sh -c '"'"'mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1464121100.3-211247575092938 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1464121100.3-211247575092938 `"'"'"''
<silence.farzy.org> PUT /var/folders/g8/tk27g4wn1r50qnhrhwm_q3xm0000gn/T/tmp7iirXA TO /root/.ansible/tmp/ansible-tmp-1464121100.3-211247575092938/stat
<silence.farzy.org> SSH: EXEC sftp -b - -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r '[silence.farzy.org]'
<silence.farzy.org> ESTABLISH SSH CONNECTION FOR USER: root
<silence.farzy.org> SSH: EXEC ssh -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r -tt silence.farzy.org '/bin/sh -c '"'"'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1464121100.3-211247575092938/stat; rm -rf "/root/.ansible/tmp/ansible-tmp-1464121100.3-211247575092938/" > /dev/null 2>&1'"'"''
<silence.farzy.org> ESTABLISH SSH CONNECTION FOR USER: root
<silence.farzy.org> SSH: EXEC ssh -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r silence.farzy.org '/bin/sh -c '"'"'mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1464121102.48-991937876545 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1464121102.48-991937876545 `"'"'"''
<silence.farzy.org> ESTABLISH SSH CONNECTION FOR USER: root
<silence.farzy.org> SSH: EXEC ssh -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r silence.farzy.org '/bin/sh -c '"'"'mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1464121102.79-78064578702232 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1464121102.79-78064578702232 `"'"'"''
<silence.farzy.org> PUT /var/folders/g8/tk27g4wn1r50qnhrhwm_q3xm0000gn/T/tmp8TJUvJ TO /root/.ansible/tmp/ansible-tmp-1464121102.79-78064578702232/file
<silence.farzy.org> SSH: EXEC sftp -b - -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r '[silence.farzy.org]'
<silence.farzy.org> ESTABLISH SSH CONNECTION FOR USER: root
<silence.farzy.org> SSH: EXEC ssh -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r -tt silence.farzy.org '/bin/sh -c '"'"'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1464121102.79-78064578702232/file'"'"''
<silence.farzy.org> PUT /tmp/accents.txt TO /root/.ansible/tmp/ansible-tmp-1464121102.48-991937876545/source
<silence.farzy.org> SSH: EXEC sftp -b - -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r '[silence.farzy.org]'
<silence.farzy.org> ESTABLISH SSH CONNECTION FOR USER: root
<silence.farzy.org> SSH: EXEC ssh -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r silence.farzy.org '/bin/sh -c '"'"'mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1464121104.9-31822415825387 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1464121104.9-31822415825387 `"'"'"''
<silence.farzy.org> PUT /var/folders/g8/tk27g4wn1r50qnhrhwm_q3xm0000gn/T/tmpMThjlL TO /root/.ansible/tmp/ansible-tmp-1464121104.9-31822415825387/copy
<silence.farzy.org> SSH: EXEC sftp -b - -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r '[silence.farzy.org]'
<silence.farzy.org> ESTABLISH SSH CONNECTION FOR USER: root
<silence.farzy.org> SSH: EXEC ssh -C -vvv -o ServerAliveInterval=60 -o ControlPersist=60s -o ForwardAgent=yes -o StrictHostKeyChecking=no -o Port=22 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=root -o ConnectTimeout=60 -o ControlPath=/tmp/.ansible/ansible-ssh-%h-%p-%r -tt silence.farzy.org '/bin/sh -c '"'"'LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1464121104.9-31822415825387/copy; rm -rf "/root/.ansible/tmp/ansible-tmp-1464121104.9-31822415825387/" > /dev/null 2>&1'"'"''
```

**Comment by abadger** [comment]
Does the task succed and it's just output that's failing?

**Comment by abadger** [comment]
I've reproduced the warning you're seeing with ansible-2.0.  It appears to only be a warning in the display of the diff, not a failing of the task.

Running it with a checkout of the 2.1 release candidate, the warning seems to have been fixed:

```
$ ansible -i 'ubuntu14,' all -m copy -a "src=/tmp/accents.txt dest=/tmp/accents.txt" --diff      
ubuntu14 | SUCCESS => {
    "changed": true, 
    "checksum": "b56a36e37922c7a3472a6efd5b3cb76e5676eb97", 
    "dest": "/tmp/accents.txt", 
    "gid": 1000, 
    "group": "badger", 
    "md5sum": "bb8a8e331cd9604c8ebf16424d620a12", 
    "mode": "0664", 
    "owner": "badger", 
    "size": 5, 
    "src": "/home/badger/.ansible/tmp/ansible-tmp-1464148098.42-68132783869136/source", 
    "state": "file", 
    "uid": 1000
}
--- before
+++ after: /tmp/accents.txt
@@ -0,0 +1 @@
+éé
```

Going to close this as fixed in 2.1.  2.1.0 should be released within the next few days.
